### PR TITLE
feat: support chunked transfer encoding

### LINKED ISSUE #648: HttpContent is encoded two times when posting data
The resulting request is correctly encoded once, but we are encoding an extra time when we calculate the content length.

### PR DESCRIPTION
**Pull request by codyoss**
Currently any time HttpRequest works with encoded data it encodes
the data twice. Once for the actual stream and once for checking
the length of the stream. Instead, when there is encoding just don't
set the content length. This will cause the underlying transports,
with a few tweaks, to use Transfer-Encoding: chunked.

Fixes #648

### CODE REVIEW

**Comment by codyoss** [comment]
Sorry for some whitespace churn. I ran the formatter for my commit, and it looks there were a handful of other files that were not complying.

File: google-http-client/src/main/java/com/google/api/client/testing/http/javanet/MockHttpURLConnection.java [hunk 1]
```diff
@@ -201,4 +201,8 @@ public String getHeaderField(String name) {
     List<String> values = headers.get(name);
     return values == null ? null : values.get(0);
   }
+
+  public int getChunckLength() {
```
Comment (chingor13) [inline google-http-client/src/main/java/com/google/api/client/testing/http/javanet/MockHttpURLConnection.java]:
```suggestion
  public int getChunkLength() {
```

**Comment by codyoss** [comment]
Sure thing.

File: google-http-client-apache-v2/src/test/java/com/google/api/client/http/apache/v2/ApacheHttpRequestTest.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
+package com.google.api.client.http.apache.v2;
```
Comment (elharo) [inline google-http-client-apache-v2/src/test/java/com/google/api/client/http/apache/v2/ApacheHttpRequestTest.java]: copyright

**Comment by codyoss** [comment]
That should be better, only formatted the files I changed. Sorry about that.

File: google-http-client/src/main/java/com/google/api/client/http/apache/ApacheHttpRequest.java [hunk 3]
```diff
@@ -54,12 +52,16 @@ public void setTimeout(int connectTimeout, int readTimeout) throws IOException {
   @Override
   public LowLevelHttpResponse execute() throws IOException {
     if (getStreamingContent() != null) {
-      Preconditions.checkArgument(request instanceof HttpEntityEnclosingRequest,
+      Preconditions.checkArgument(
+          request instanceof HttpEntityEnclosingRequest,
```
Comment (elharo) [inline google-http-client/src/main/java/com/google/api/client/http/apache/ApacheHttpRequest.java]: request is not an argument here. This should be checkState

File: google-http-client-apache-v2/src/main/java/com/google/api/client/http/apache/v2/ApacheHttpRequest.java [hunk 4]
```diff
@@ -52,19 +51,22 @@ public void addHeader(String name, String value) {
...
-    requestConfig.setConnectTimeout(connectTimeout)
-        .setSocketTimeout(readTimeout);
+    requestConfig.setConnectTimeout(connectTimeout).setSocketTimeout(readTimeout);
   }
 
   @Override
   public LowLevelHttpResponse execute() throws IOException {
     if (getStreamingContent() != null) {
-      Preconditions.checkArgument(request instanceof HttpEntityEnclosingRequest,
+      Preconditions.checkArgument(
```
Comment (elharo) [inline google-http-client-apache-v2/src/main/java/com/google/api/client/http/apache/v2/ApacheHttpRequest.java]: checkState

File: google-http-client/src/test/java/com/google/api/client/http/javanet/NetHttpRequestTest.java [hunk 5]
```diff
@@ -203,4 +203,34 @@ public void close() throws IOException {
...
+    MockHttpURLConnection connection = new MockHttpURLConnection(new URL(HttpTesting.SIMPLE_URL));
+    connection.setRequestMethod("POST");
+    NetHttpRequest request = new NetHttpRequest(connection);
+    InputStream is = NetHttpRequestTest.class.getClassLoader().getResourceAsStream("file.txt");
+    HttpContent content = new InputStreamContent("text/plain", is);
+    request.setStreamingContent(content);
+    request.setContentEncoding("gzip");
+    request.execute();
+
+    assertEquals(connection.getChunkLength(), 4096);
```
Comment (elharo) [inline google-http-client/src/test/java/com/google/api/client/http/javanet/NetHttpRequestTest.java]: expected value comes first so assertEquals(4096, connection.getChunkLength());

**Pull request merged**